### PR TITLE
Fix: add missing python wrappers for MolDraw2DQt

### PR DESCRIPTION
**Pull request by greglandrum**
This is a bit tricky since PyQt is wrapped using sip and we need boost::python wrappers.
Fortunately sip provides a function which allows us to get the raw pointer behind one of its objects and we can pass that directly to C++.
There's another bit of fun around lifetime management of the objects - trying to GC the QPainter object before we've removed the MolDraw2DQt object results in a seg fault.
This handles both problems by providing a python convenience function which creates a MolDraw2DQt object from a QPainter and sets up the lifetime management properly.

### CODE REVIEW

**Comment by greglandrum** [comment]
@ZontaNicola: would appreciate your feedback on this one. 
There's not a lot there, but if you have suggestions about ways to handle the PyQt<->Boost thing more cleanly I'd welcome them.

File: Code/GraphMol/MolDraw2D/Wrap/rdMolDraw2D.cpp [hunk 1]
```diff
@@ -543,6 +547,17 @@ void setDrawerColour(RDKit::MolDraw2D &self, python::tuple tpl) {
   self.setColour(pyTupleToDrawColour(tpl));
 }
 
+#ifdef RDK_BUILD_QT_SUPPORT
+MolDraw2DQt *moldrawFromQPainter(int width, int height, unsigned long ptr,
```
Comment (ZontaNicola) [inline Code/GraphMol/MolDraw2D/Wrap/rdMolDraw2D.cpp:551]: why are we not passing ptr as QPointer*?

File: Code/GraphMol/MolDraw2D/Wrap/testMolDraw2DQt.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+  def setUp(self):
+    pass
+
+  def testSIPBasics(self):
+    m = Chem.MolFromSmiles('c1ccccc1O')
+    Draw.PrepareMolForDrawing(m)
+    qimg = QImage(250, 200, QImage.Format_RGB32)
+    qptr = QPainter(qimg)
+    qptr._img = qimg
```
Comment (ZontaNicola) [inline Code/GraphMol/MolDraw2D/Wrap/testMolDraw2DQt.py:22]: what does this line do?

File: Code/GraphMol/MolDraw2D/Wrap/rdMolDraw2D.cpp (hunk 1 above)
Comment (greglandrum) [reply Code/GraphMol/MolDraw2D/Wrap/rdMolDraw2D.cpp:551]: On the Python side we just have the PyQt wrapped object and we can't get the QPointer* from that. The best we can do, as far as I can tell, is get the pointer as an int using sip.unwrapobject and pass that ,

File: Code/GraphMol/MolDraw2D/Wrap/testMolDraw2DQt.py (hunk 2 above)
Comment (greglandrum) [reply Code/GraphMol/MolDraw2D/Wrap/testMolDraw2DQt.py:22]: lifetime management for qimg: it ensures that qimg won't be deleted while qptr is still alive

**Comment by greglandrum** [comment]
FWIW: the test failures here are almost certainly because I have the linux CI build setup to use the system Qt and PyQt5 from conda. That's not a recipe for success, so I'll need to do some more work on the linux-ci configuration. This shouldn't impact the review of the PR though

**Comment by ricrogz** [comment]
I'm far from being an expert in Qt, PyQt, or python boost, but even though the unwrapping/casting method doesn't look very nice to me, it seems to be what other people are doing (at least from what I have googled around to see if other people had similar problems). The object's "lifetime management" workaround looks a bit hackish, but I guess it also comes from the same issue: PyQt and boost::python don't seem to be designed to interact between themselves.

Regarding the tests, I have downloaded this patch and tried the tests locally, and found some problems: when I import `rdkit.Chem.Draw`, `MolDraw2DFromQPainter` isn't defined, and so tests `testBasics` and `testMemory2` fail with errors. `MolDraw2DFromQPainter_` (note the final underscore) is defined and working, though.

Apparently, I'm having some problems to import `sip`, which explains why the function isn't defined. If I change __init__.py to use `from PyQt5 import sip` instead of `import sip`, then everything works fine. I think it might make sense to update this code to use sip from PyQt5, so that we are guaranteed to use the same version that was used to generate the Qt wrappers. Maybe this could also explain the segv Greg observed ?

File: Code/GraphMol/MolDraw2D/Wrap/testMolDraw2DQt.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
+from rdkit import RDConfig
+import unittest
+from rdkit import Chem
+from rdkit.Chem import Draw, AllChem, rdDepictor
+from rdkit.Chem.Draw import rdMolDraw2D
+
+import sys
+try:
+  from PyQt5.Qt import *
+  import sip
```
Comment (ricrogz) [inline Code/GraphMol/MolDraw2D/Wrap/testMolDraw2DQt.py:10]:
```suggestion
  from PyQt5 import sip
```

File: rdkit/Chem/Draw/__init__.py [hunk 4]
```diff
@@ -149,6 +149,29 @@ def _legacyMolToImage(mol, size, kekulize, wedgeBonds, fitImage, options, canvas
     return img
 
 
+if hasattr(rdMolDraw2D, 'MolDraw2DQt'):
+  try:
+    import sip
```
Comment (ricrogz) [inline rdkit/Chem/Draw/__init__.py:154]:
```suggestion
    from PyQt5 import sip
```

**Review by ricrogz — COMMENTED** [review]
My suggestion to use sip from PyQt5, instead of the separate package.

**Comment by greglandrum** [comment]
> My suggestion to use sip from PyQt5, instead of the separate package.

Ok, I made those changes, but I don't think that's actually going to help.


For conda installs of PyQt5 there is no PyQt5.sip module. Installing the pyqt conda package gets you a standalone sip package (this is what I've been using).
```
(rdkit_build) root@3ce870abb14b:/src/rdkit/build# conda install PyQt 
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/rdkit_build

  added / updated specs:
    - pyqt


The following NEW packages will be INSTALLED:

  pyqt               pkgs/main/linux-64::pyqt-5.9.2-py38h05f1152_4
  sip                pkgs/main/linux-64::sip-4.19.13-py38he6710b0_0


Proceed ([y]/n)? 

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
(rdkit_build) root@3ce870abb14b:/src/rdkit/build# ipython
Python 3.8.5 (default, Sep  4 2020, 07:30:14) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.19.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from PyQt5 import sip
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-1-d9a10bac1b11> in <module>
----> 1 from PyQt5 import sip

ImportError: cannot import name 'sip' from 'PyQt5' (/opt/conda/envs/rdkit_build/lib/python3.8/site-packages/PyQt5/__init__.py)

```

**Comment by ricrogz** [comment]
> For conda installs of PyQt5 there is no PyQt5.sip module. Installing the pyqt conda package gets you a standalone sip package (this is what I've been using).

Oh, I see. I'm using "vanilla" Python, and it seems there's separate packages for PyQt5, PyQt5-sip and sip (I have more or less the same on 2 different linux machines and a Mac, all of them with "vanilla" Python):

Package | version | PyPI page
--- | --- | ---
PyQt5 | 5.12.2 | https://pypi.org/project/PyQt5/
PyQt5-sip | 4.19.17 | https://pypi.org/project/PyQt5-sip/
sip | 5.5.0 | https://pypi.org/project/sip/

`PyQt5-sip` and `sip` versions are quite different, and my `PyQt5-sip` version seems to be close to conda's `sip`. I'd say that while PyPI's `sip` is the last available release, PyPI's `PyQt5-sip` and conda's `sip` are actually the specific sip version that was used to build PyQt5.

That's a bit messy, but I think we could use something like:

```
try:
    from PyQt5.Qt import *
except ImportError:
    # If we can't find Qt, there's nothing we can do
    sip = None
else:
    try:
        # Prefer the PyQt5-bundled sip
        from PyQt5 import sip
    except ImportError:
        # No bundled sip, try the standalone package
        try:
            import sip
        except ImportError:
            # No sip at all
            sip = None
```

**Pull request merged**